README.md
# aurelia-loader-default

[![npm Version](https://img.shields.io/npm/v/aurelia-loader-default.svg)](https://www.npmjs.com/package/aurelia-loader-default)
[![ZenHub](https://raw.githubusercontent.com/ZenHubIO/support/master/zenhub-badge.png)](https://zenhub.io)
[![Join the chat at https://gitter.im/aurelia/discuss](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/aurelia/discuss?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![CircleCI](https://circleci.com/gh/aurelia/loader-default.svg?style=shield)](https://circleci.com/gh/aurelia/loader-default)

This library is part of the [Aurelia](http://www.aurelia.io/) platform and contains an implementation of the loader interface which works with system.js and all require-based loader APIs.

> To keep up to date on [Aurelia](http://www.aurelia.io/), please visit and subscribe to [the official blog](http://blog.aurelia.io/) and [our email list](http://eepurl.com/ces50j). We also invite you to [follow us on twitter](https://twitter.com/aureliaeffect). If you have questions look around our [Discourse forums](https://discourse.aurelia.io/), chat in our [community on Gitter](https://gitter.im/aurelia/discuss) or use [stack overflow](http://stackoverflow.com/search?q=aurelia). Documentation can be found [in our developer hub](http://aurelia.io/docs). If you would like to have deeper insight into our development process, please install the [ZenHub](https://zenhub.io) Chrome or Firefox Extension and visit any of our repository's boards.

## Platform Support

This library can be used in the **browser** only.

## Building The Code

To build the code, follow these steps.

1. Ensure that [NodeJS](http://nodejs.org/) is installed. This provides the platform on which the build tooling runs.
2. From the project folder, execute the following command:

  ```shell
  npm install
  ```
3. Ensure that [Gulp](http://gulpjs.com/) is installed. If you need to install it, use the following command:

  ```shell
  npm install -g gulp
  ```
4. To build the code, you can now run:

  ```shell
  gulp build
  ```
5. You will find the compiled code in the `dist` folder, available in three module formats: AMD, CommonJS and ES6.

6. See `gulpfile.js` for other tasks related to generating the docs and linting.

## Running The Tests

To run the unit tests, first ensure that you have followed the steps above in order to install all dependencies and successfully build the library. Once you have done that, proceed with these additional steps:

1. Ensure that the [Karma](http://karma-runner.github.io/) CLI is installed. If you need to install it, use the following command:

  ```shell
  npm install -g karma-cli
  ```
2. Ensure that [jspm](http://jspm.io/) is installed. If you need to install it, use the following commnand:

  ```shell
  npm install -g jspm
  ```
3. Install the client-side dependencies with jspm:

  ```shell
  jspm install
  ```

4. You can now run the tests with this command:

  ```shell
  karma start
  ```


doc/CHANGELOG.md
<a name="1.2.2"></a>
## [1.2.2](https://github.com/aurelia/loader-default/compare/1.2.0...1.2.2) (2019-03-27)



<a name="1.2.1"></a>
## [1.2.1](https://github.com/aurelia/loader-default/compare/1.2.0...1.2.1) (2019-02-04)



<a name="1.2.0"></a>
# [1.2.0](https://github.com/aurelia/loader-default/compare/1.1.0...1.2.0) (2019-01-18)


### Bug Fixes

* do not reject unsupported AMD loader ([846befe](https://github.com/aurelia/loader-default/commit/846befe))


### Features

* add support of dumber-module-loader ([1d9433f](https://github.com/aurelia/loader-default/commit/1d9433f))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/aurelia/loader-default/compare/1.0.4...1.1.0) (2018-12-01)


### Bug Fixes

* remove unnecessary inline SystemJS text plugin ([8c9a183](https://github.com/aurelia/loader-default/commit/8c9a183)), closes [aurelia/i18n#289](https://github.com/aurelia/i18n/issues/289)


### Features

* **loader:** Added support for alameda amd loader ([9b4bb40](https://github.com/aurelia/loader-default/commit/9b4bb40))



<a name="1.0.4"></a>
## [1.0.4](https://github.com/aurelia/loader-default/compare/1.0.3...1.0.4) (2018-03-29)


### Bug Fixes

* fix commonjs build error in requirejs env ([04f6d60](https://github.com/aurelia/loader-default/commit/04f6d60)), closes [#47](https://github.com/aurelia/loader-default/issues/47)
* improve compatibility with other AMD implementations. ([0f8b5e8](https://github.com/aurelia/loader-default/commit/0f8b5e8)), closes [#48](https://github.com/aurelia/loader-default/issues/48)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/aurelia/loader-default/compare/1.0.2...v1.0.3) (2017-10-01)


### Bug Fixes

* **index:** use .keys instead of .entries close [#43](https://github.com/aurelia/loader-default/issues/43) ([49948a1](https://github.com/aurelia/loader-default/commit/49948a1)), closes [#43](https://github.com/aurelia/loader-default/issues/43)


### Performance Improvements

* **index:** single array index for system key ([f5e36b1](https://github.com/aurelia/loader-default/commit/f5e36b1))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/aurelia/loader-default/compare/1.0.1...v1.0.2) (2017-03-23)


### Bug Fixes

* **index:** fix PLATFORM.eachModule for SystemJS[@0](https://github.com/0).20.x ([7c2c925](https://github.com/aurelia/loader-default/commit/7c2c925))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/aurelia/loader-default/compare/1.0.0...v1.0.1) (2017-02-21)


### Bug Fixes

* **DefaultLoader:** use config API for map ([a4607b9](https://github.com/aurelia/loader-default/commit/a4607b9))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/aurelia/loader-default/compare/1.0.0-rc.1.0.0...v1.0.0) (2016-07-27)



<a name="1.0.0-rc.1.0.0"></a>
# [1.0.0-rc.1.0.0](https://github.com/aurelia/loader-default/compare/1.0.0-beta.2.0.2...v1.0.0-rc.1.0.0) (2016-06-22)



### 1.0.0-beta.1.2.2 (2016-05-10)


### 1.0.0-beta.1.2.1 (2016-04-13)


#### Bug Fixes

* **index:** ensure correct cache lookup ([b32ff61a](http://github.com/aurelia/loader-default/commit/b32ff61a5fc00eb995bb008d7c1f83fdffa71d7d))


### 1.0.0-beta.1.2.0(2016-03-22)

#### Bug Fixes

* **DefaultLoader:** do not map name to address in requirejs mode ([30016b6e](http://github.com/aurelia/loader-default/commit/30016b6ef904228a582863e83b49e8f3c0556e63))

### 1.0.0-beta.1.1.3 (2016-02-25)


#### Bug Fixes

* **loader:** alias the loader ([70b7cb83](http://github.com/aurelia/loader-default/commit/70b7cb83a5f78131cdb66bc29398c24fd243b2b1))


### 1.0.0-beta.1.1.2 (2016-02-08)


#### Bug Fixes

* **index:** properly normalize for systemjs ([aa3c4813](http://github.com/aurelia/loader-default/commit/aa3c48136a829ff1137d8ecb5f367d7b5e05cf0d))


### 1.0.0-beta.1.1.0 (2016-01-29)


#### Features

* **all:** update jspm meta; core-js; aurelia deps ([7777dfcc](http://github.com/aurelia/loader-default/commit/7777dfcca39fb610b7dd66bad868133b2d0c9962))


### 1.0.0-beta.1.0.2 (2016-01-08)


#### Features

* **all:** remove internal normalizeSync use and add normalize impl ([3c323466](http://github.com/aurelia/loader-default/commit/3c323466f461a9831675aa9421968b2727c50164))


## 1.0.0-beta.1.0.1 (2015-11-17)


#### Bug Fixes

* **index:** firefox can die on some module getter accessors ([97dee32c](http://github.com/aurelia/loader-default/commit/97dee32c0df0359ec722ea852ab5529e98ae34be))


### 1.0.0-beta.1 (2015-11-16)


## 0.12.0 (2015-11-10)


#### Bug Fixes

* **text-template-loader:** update to latest template entry api ([9fd102dc](http://github.com/aurelia/loader-default/commit/9fd102dc55af408b0393844442171a7a9b31ca93))


### 0.11.2 (2015-10-13)


#### Bug Fixes

* **all:**
  * remove web components and html imports ([5e40af0f](http://github.com/aurelia/loader-default/commit/5e40af0f22ccf90b7f959fb770cbf2cd5a9f7c8e))
  * improve the template loader interface and html imports lookup ([dec84529](http://github.com/aurelia/loader-default/commit/dec84529b0dc6174c5daa9609ecaf2304515886f))
  * differentiate between real, polyfill and fake system ([09767418](http://github.com/aurelia/loader-default/commit/09767418d730aac3e772c0de954b18159673ee68))
  * update compiler ([11498313](http://github.com/aurelia/loader-default/commit/11498313b369c5752f69190ddecfccaaae7ea32b))
* **build:**
  * update linting, testing and tools ([9b8935a2](http://github.com/aurelia/loader-default/commit/9b8935a297c2b7807acf7be429fe6a528c9ceef8))
  * add missing bower bump ([a3e96402](http://github.com/aurelia/loader-default/commit/a3e96402dbdd48da9904fe228c4ddc5b9d32ef97))
* **index:**
  * sometimes undefined modules list in eachModule iterator ([fe4adddc](http://github.com/aurelia/loader-default/commit/fe4adddce7517c282011f6728b448fbf4d598ef9))
  * incorrect variable name in require-based default loader code ([4cdf9df5](http://github.com/aurelia/loader-default/commit/4cdf9df55d4eecedd5d43fddc9b41871b858818e), closes [#27](http://github.com/aurelia/loader-default/issues/27))

test/loader.spec.js
import './setup';
import {DefaultLoader} from '../src/index';

describe('the system.js loader', () => {
  describe('instance method', () => {
    var loader;
    beforeEach(() => {
      System.config({
        "paths": {
          "*": "test/fixtures/*.js"
        }
      });

      loader = new DefaultLoader();
    });

    describe('loadModule', () => {
      it('should load a module', (done) => {
        loader.loadModule('baseModule')
          .then((result) => {
            expect(result).toEqual(jasmine.any(Object));
            expect(result.default).toEqual(jasmine.any(Function));
            expect(result.default()).toEqual("baseModule hello");
          })
          .catch((reason) => expect(false).toBeTruthy(reason))
          .then(done);
      });

      it("should fail if the module doesn't exist", (done) => {
        loader.loadModule('notHere')
          .then(result => expect(false).toBeTruthy('No module should have been found'))
          .catch(reason => expect(reason).toEqual(jasmine.any(Error)))
          .then(done);
      });

      it("should cache any repeated module calls", (done) => {
        let importSpy = spyOn(System, 'import').and.callThrough();

        //If both calls are done right after each other then the loader won't cache.
        //Perhaps this is a potential optimization
        loader.loadModule('baseModule').then(() => {
          return loader.loadModule('baseModule').then(() => {
            expect(importSpy.calls.count()).toEqual(1);
          });
        })
          .catch((reason) => expect(false).toBeTruthy(reason))
          .then(done);
      });
    });

    describe('loadAllModules()', () => {

      it("will return when all modules are loaded", (done) => {
        loader.loadAllModules(['baseModule', 'moduleBase/rebasedModule'])
          .then((results) => {
            expect(results.length).toBe(2);
            for (let result of results) {
              expect(result).toEqual(jasmine.any(Object));
              expect(result.default).toEqual(jasmine.any(Function));
              expect(result.default()).toEqual(jasmine.any(String));
            }
          })
          .catch((reason) => expect(false).toBeTruthy(reason))
          .then(done);
      });

      it("will fail if any modules fail to load", (done) => {
        loader.loadAllModules(['baseModule', 'doesntExist'])
          .then(() => expect(false).toBeTruthy('No Modules should have loaded'))
          .catch((reason) => expect(reason).toEqual(jasmine.any(Error)))
          .then(done);
      });
    });

    describe('loadTemplate()', () => {

      it('will load a template', (done) => {
        loader.loadTemplate('baseTemplate.html')
        .then((result) => {
            expect(result.template.innerHTML).toBe("<h1>I am the base template</h1>");
          })
        .catch((reason) => expect(false).toBeTruthy(reason))
        .then(done);
      });

      //At this point I can't test for failing template loads as they never resolve.
      //This is an issue with tbe base loader.
    });

  });
});


test/fixtures/text.js
/*
  Text plugin
*/
exports.translate = function(load) {
  if (this.builder && this.transpiler) {
    load.metadata.format = 'esm';
    return 'exp' + 'ort var __useDefault = ' + JSON.stringify(load.source) + '; exp' + 'ort default __useDefault;';
  }

  load.metadata.format = 'amd';
  return 'def' + 'ine(function() {\nreturn ' + JSON.stringify(load.source) + ';\n});';
};


test/fixtures/baseModule.js
/**
 * Created by Andrew on 2015-02-16.
 */
export class baseModule{
  static message(){
    return "I am the baseModule message";
  }
}


config.js
System.config({
  defaultJSExtensions: true,
  transpiler: "babel",
  babelOptions: {
    "optional": [
      "runtime",
      "optimisation.modules.system"
    ]
  },
  paths: {
    "github:*": "jspm_packages/github/*",
    "aurelia-loader-default/*": "dist/*",
    "npm:*": "jspm_packages/npm/*"
  },

  map: {
    "aurelia-loader": "npm:aurelia-loader@1.0.0",
    "aurelia-metadata": "npm:aurelia-metadata@1.0.0",
    "aurelia-pal": "npm:aurelia-pal@1.0.0",
    "aurelia-pal-browser": "npm:aurelia-pal-browser@1.0.0",
    "babel": "npm:babel-core@5.8.38",
    "babel-runtime": "npm:babel-runtime@5.8.38",
    "core-js": "npm:core-js@2.4.1",
    "github:jspm/nodelibs-assert@0.1.0": {
      "assert": "npm:assert@1.4.1"
    },
    "github:jspm/nodelibs-buffer@0.1.0": {
      "buffer": "npm:buffer@3.6.0"
    },
    "github:jspm/nodelibs-path@0.1.0": {
      "path-browserify": "npm:path-browserify@0.0.0"
    },
    "github:jspm/nodelibs-process@0.1.2": {
      "process": "npm:process@0.11.6"
    },
    "github:jspm/nodelibs-util@0.1.0": {
      "util": "npm:util@0.10.3"
    },
    "github:jspm/nodelibs-vm@0.1.0": {
      "vm-browserify": "npm:vm-browserify@0.0.4"
    },
    "npm:assert@1.4.1": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "buffer": "github:jspm/nodelibs-buffer@0.1.0",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "util": "npm:util@0.10.3"
    },
    "npm:aurelia-loader@1.0.0": {
      "aurelia-metadata": "npm:aurelia-metadata@1.0.0",
      "aurelia-path": "npm:aurelia-path@1.0.0"
    },
    "npm:aurelia-metadata@1.0.0": {
      "aurelia-pal": "npm:aurelia-pal@1.0.0"
    },
    "npm:aurelia-pal-browser@1.0.0": {
      "aurelia-pal": "npm:aurelia-pal@1.0.0"
    },
    "npm:babel-runtime@5.8.38": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:buffer@3.6.0": {
      "base64-js": "npm:base64-js@0.0.8",
      "child_process": "github:jspm/nodelibs-child_process@0.1.0",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "ieee754": "npm:ieee754@1.1.6",
      "isarray": "npm:isarray@1.0.0",
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:core-js@2.4.1": {
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "path": "github:jspm/nodelibs-path@0.1.0",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "systemjs-json": "github:systemjs/plugin-json@0.1.2"
    },
    "npm:inherits@2.0.1": {
      "util": "github:jspm/nodelibs-util@0.1.0"
    },
    "npm:path-browserify@0.0.0": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:process@0.11.6": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "vm": "github:jspm/nodelibs-vm@0.1.0"
    },
    "npm:util@0.10.3": {
      "inherits": "npm:inherits@2.0.1",
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:vm-browserify@0.0.4": {
      "indexof": "npm:indexof@0.0.1"
    }
  }
});


src/index.js
/*eslint dot-notation:0*/
import {Origin} from 'aurelia-metadata';
import {TemplateRegistryEntry, Loader} from 'aurelia-loader';
import {TextTemplateLoader} from './text-template-loader';
import {PLATFORM} from 'aurelia-pal';

function ensureOriginOnExports(executed, name) {
  let target = executed;
  let key;
  let exportedValue;

  if (target.__useDefault) {
    target = target['default'];
  }

  Origin.set(target, new Origin(name, 'default'));

  for (key in target) {
    exportedValue = target[key];

    if (typeof exportedValue === 'function') {
      Origin.set(exportedValue, new Origin(name, key));
    }
  }

  return executed;
}

/**
* Represents a template loader.
*/
interface TemplateLoader {
  /**
  * Loads a template.
  * @param loader The loader that is requesting the template load.
  * @param entry The TemplateRegistryEntry to load and populate with a template.
  * @return A promise which resolves when the TemplateRegistryEntry is loaded with a template.
  */
  loadTemplate(loader: Loader, entry: TemplateRegistryEntry): Promise<any>;
}

/**
* A default implementation of the Loader abstraction which works with SystemJS, RequireJS and Dojo Loader.
*/
export class DefaultLoader extends Loader {
  /**
  * The name of the underlying native loader plugin used to load text.
  */
  textPluginName: string = 'text';

  /**
  * Creates an instance of the DefaultLoader.
  */
  constructor() {
    super();

    this.moduleRegistry = Object.create(null);
    this.useTemplateLoader(new TextTemplateLoader());

    let that = this;

    this.addPlugin('template-registry-entry', {
      'fetch': function(address) {
        let entry = that.getOrCreateTemplateRegistryEntry(address);
        return entry.templateIsLoaded ? entry : that.templateLoader.loadTemplate(that, entry).then(x => entry);
      }
    });
  }

  /**
  * Instructs the loader to use a specific TemplateLoader instance for loading templates
  * @param templateLoader The instance of TemplateLoader to use for loading templates.
  */
  useTemplateLoader(templateLoader: TemplateLoader): void {
    this.templateLoader = templateLoader;
  }

  /**
  * Loads a collection of modules.
  * @param ids The set of module ids to load.
  * @return A Promise for an array of loaded modules.
  */
  loadAllModules(ids: string[]): Promise<any[]> {
    let loads = [];

    for (let i = 0, ii = ids.length; i < ii; ++i) {
      loads.push(this.loadModule(ids[i]));
    }

    return Promise.all(loads);
  }

  /**
  * Loads a template.
  * @param url The url of the template to load.
  * @return A Promise for a TemplateRegistryEntry containing the template.
  */
  loadTemplate(url: string): Promise<TemplateRegistryEntry> {
    return this._import(this.applyPluginToUrl(url, 'template-registry-entry'));
  }

  /**
  * Loads a text-based resource.
  * @param url The url of the text file to load.
  * @return A Promise for text content.
  */
  loadText(url: string): Promise<string> {
    return this._import(this.applyPluginToUrl(url, this.textPluginName)).then(textOrModule => {
      if (typeof textOrModule === 'string') {
        return textOrModule;
      }

      return textOrModule['default'];
    });
  }
}

PLATFORM.Loader = DefaultLoader;

if (!PLATFORM.global.System || !PLATFORM.global.System.import) {
  if (PLATFORM.global.requirejs) {
    let getDefined;
    if (typeof PLATFORM.global.requirejs.s === 'object') {
      // Support for requirejs/requirejs
      getDefined = () => PLATFORM.global.requirejs.s.contexts._.defined;
    } else if (typeof PLATFORM.global.requirejs.contexts === 'object') {
      // Support for requirejs/alameda
      getDefined = () => PLATFORM.global.requirejs.contexts._.defined;
    } else if (typeof PLATFORM.global.requirejs.definedValues === 'function') {
      // Support for dumberjs/dumber-module-loader
      getDefined = () => PLATFORM.global.requirejs.definedValues();
    } else {
      // skip any unknown AMD loader
      getDefined = () => ({});
    }
    PLATFORM.eachModule = function(callback) {
      const defined = getDefined();
      for (let key in defined) {
        try {
          if (callback(key, defined[key])) return;
        } catch (e) {}
      }
    };
  } else {
    PLATFORM.eachModule = function(callback) {};
  }

  DefaultLoader.prototype._import = function(moduleId) {
    return new Promise((resolve, reject) => {
      PLATFORM.global.require([moduleId], resolve, reject);
    });
  };

  DefaultLoader.prototype.loadModule = function(id) {
    let existing = this.moduleRegistry[id];
    if (existing !== undefined) {
      return Promise.resolve(existing);
    }

    return new Promise((resolve, reject) => {
      PLATFORM.global.require([id], m => {
        this.moduleRegistry[id] = m;
        resolve(ensureOriginOnExports(m, id));
      }, reject);
    });
  };

  DefaultLoader.prototype.map = function(id, source) {};

  DefaultLoader.prototype.normalize = function(moduleId, relativeTo) {
    return Promise.resolve(moduleId);
  };

  DefaultLoader.prototype.normalizeSync = function(moduleId, relativeTo) {
    return moduleId;
  };

  DefaultLoader.prototype.applyPluginToUrl = function(url, pluginName) {
    return `${pluginName}!${url}`;
  };

  DefaultLoader.prototype.addPlugin = function(pluginName, implementation) {
    let nonAnonDefine = define;
    nonAnonDefine(pluginName, [], {
      'load': function(name, req, onload) {
        let result = implementation.fetch(name);
        Promise.resolve(result).then(onload);
      }
    });
  };
} else {
  PLATFORM.eachModule = function(callback) {
    if (System.registry) { // SystemJS >= 0.20.x
      const keys = Array.from(System.registry.keys());
      for (let i = 0; i < keys.length; i++) {
        try {
          let key = keys[i];
          if (callback(key, System.registry.get(key))) { return; }
        } catch (e) {}
      }
      return;
    }

    // SystemJS < 0.20.x
    let modules = System._loader.modules;

    for (let key in modules) {
      try {
        if (callback(key, modules[key].module)) return;
      } catch (e) {}
    }
  };

  DefaultLoader.prototype._import = function(moduleId) {
    return System.import(moduleId);
  };

  DefaultLoader.prototype.loadModule = function(id) {
    return System.normalize(id).then(newId => {
      let existing = this.moduleRegistry[newId];
      if (existing !== undefined) {
        return Promise.resolve(existing);
      }

      return System.import(newId).then(m => {
        this.moduleRegistry[newId] = m;
        return ensureOriginOnExports(m, newId);
      });
    });
  };

  DefaultLoader.prototype.map = function(id, source) {
    System.config({ map: { [id]: source } });
  };

  DefaultLoader.prototype.normalizeSync = function(moduleId, relativeTo) {
    return System.normalizeSync(moduleId, relativeTo);
  };

  DefaultLoader.prototype.normalize = function(moduleId, relativeTo) {
    return System.normalize(moduleId, relativeTo);
  };

  DefaultLoader.prototype.applyPluginToUrl = function(url, pluginName) {
    return `${url}!${pluginName}`;
  };

  DefaultLoader.prototype.addPlugin = function(pluginName, implementation) {
    System.set(pluginName, System.newModule({
      'fetch': function(load, _fetch) {
        let result = implementation.fetch(load.address);
        return Promise.resolve(result).then(x => {
          load.metadata.result = x;
          return '';
        });
      },
      'instantiate': function(load) {
        return load.metadata.result;
      }
    }));
  };
}


src/html-import-template-loader.js
import {TemplateRegistryEntry, Loader} from 'aurelia-loader';

export class HTMLImportTemplateLoader {
  constructor() {
    this.hasTemplateElement = ('content' in document.createElement('template'));
    this.needsBundleCheck = true;
    this.onBundleReady = null;
  }

  loadTemplate(loader: Loader, entry: TemplateRegistryEntry): Promise<any> {
    return this._tryFindTemplateInBundle(entry).then(found => {
      return found ? entry : this._importDocument(entry).then(doc => this._findTemplate(doc, entry));
    });
  }

  _tryFindTemplateInBundle(entry: TemplateRegistryEntry): Promise<boolean> {
    if (this.bundle) {
      return this._tryGetTemplateFromBundle(entry);
    } else if (this.onBundleReady) {
      return this.onBundleReady.then(() => this._tryGetTemplateFromBundle(entry));
    } else if (this.needsBundleCheck) {
      if (!('import' in document.createElement('link'))) {
        return System.normalize('aurelia-loader-default').then(name => {
          return System.import('webcomponentsjs/HTMLImports.min', name).then(() => {
            return this._loadBundle(entry);
          });
        });
      }

      return this._loadBundle(entry);
    }

    return Promise.resolve(false);
  }

  _loadBundle(entry) {
    let bundleLink = document.querySelector('link[aurelia-view-bundle]');
    this.needsBundleCheck = false;

    if (bundleLink) {
      this.onBundleReady = this._importBundle(bundleLink).then(doc => {
        this._normalizeTemplateIds(doc);
        this.bundle = doc;
        this.onBundleReady = null;
      });

      return this.onBundleReady.then(() => this._tryGetTemplateFromBundle(entry));
    }

    return Promise.resolve(false);
  }

  _importDocument(entry) {
    return new Promise((resolve, reject) => {
      let frag = document.createDocumentFragment();
      let link = document.createElement('link');

      link.rel = 'import';
      link.href = entry.address;
      frag.appendChild(link);

      this._importElements(frag, link, () => resolve(link.import));
    });
  }

  _findTemplate(doc, entry) {
    if (!this.hasTemplateElement) {
      HTMLTemplateElement.bootstrap(doc);
    }

    let template = doc.getElementsByTagName('template')[0];

    if (!template) {
      throw new Error(`There was no template element found in '${entry.address}'.`);
    }

    entry.setTemplate(template);
  }

  _tryGetTemplateFromBundle(entry) {
    let found = this.bundle.getElementById(entry.address);

    if (found) {
      entry.setTemplate(found);
      return Promise.resolve(true);
    }

    return Promise.resolve(false);
  }

  _importBundle(link) {
    return new Promise((resolve, reject) => {
      if (link.import) {
        if (!this.hasTemplateElement) {
          HTMLTemplateElement.bootstrap(link.import);
        }

        resolve(link.import);
      } else {
        this._importElements(null, link, () => {
          if (!this.hasTemplateElement) {
            HTMLTemplateElement.bootstrap(link.import);
          }

          resolve(link.import);
        });
      }
    });
  }

  _normalizeTemplateIds(doc) {
    let templates = doc.getElementsByTagName('template');

    for (let i = 0, ii = templates.length; i < ii; ++i) {
      let current = templates[i];
      current.setAttribute('id', System.normalizeSync(current.getAttribute('id')));
    }
  }

  _importElements(frag, link, callback) {
    if (frag) {
      document.head.appendChild(frag);
    }

    if (window.Polymer && Polymer.whenReady ) {
      Polymer.whenReady(callback);
    } else {
      link.addEventListener('load', callback);
    }
  }
}


karma.conf.js
// Karma configuration
// Generated on Fri Dec 05 2014 16:49:29 GMT-0500 (EST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jspm', 'jasmine'],

    jspm: {
      // Edit this to your needs
      loadFiles: ['test/**/*.js'],
      serveFiles : ['src/**/*.js']
    },


    // list of files / patterns to load in the browser
    files: [
      {pattern: 'test/fixtures/**/*.html', included: false}
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/**/*.js': ['babel'],
      'src/**/*.js': ['babel']
    },
    'babelPreprocessor': {
      options: {
        sourceMap: 'inline',
        presets: [ 'es2015-loose', 'stage-1'],
        plugins: [
          'syntax-flow',
          'transform-decorators-legacy',
          'transform-flow-strip-types'
        ]
      }
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};
